### ISSUE TITLE
Server crashing due to FAWE deadlock

### ISSUE THREAD

**Issue by Weark** [Imanity-Software/ImanitySpigot3-git#137]
### Stack trace

https://paste.imanity.dev/HjtAVT7

### Plugin list

Plugins (86): NameMC-API, OldEnchanting, GiveAll, PlaceholderAPI, FriendsAPIForPartyAndFriends, LeaderHeadsRevamped, CommandTimer, DisableExpOrb, PartyAndFriendsGUI, Spigot-Party-API-PAF, PlotBorder, HealthBar, HeadHunter, ClearLag, AutoMessage, UltraMySQLPAPIBridge, spark, Votifier, IronElevators, KiteBoard, VoidSpawn, AntiDrop, FastAsyncWorldEdit, DBVerifierLinker, UnbreakableAnvils, ProVotes, CrazyCrates, Mines, KeepDeathDrop, WorldEdit, WorldGuard, BanItem, PixelVip, BetterSocial, ChestShop, ChestShopNotifier, TheNewEconomy, DeluxeCombat, DeluxeCombat_Addon_WG, PlotSquared, antiRedstoneClock, BattlePass, HexagonClearPlot, Multiverse-Core, Insights, Multiverse-Patches, ServerNPC, DailyBonus, SkinsRestorer, Vulcan, TradePlus, ViaVersion, LuckPerms, SpigotGuard, ProtocolLib, RoyaleEconomy, zTournament, Vault, PengAPI, Pl-Hide-Pro, PengChat, PengFullPvP, MyCommand, MysqlPlayerDataBridge, Essentials, PlayerWarps, VKAutoPickup, EZRanksLite, GadgetsMenu, HeadDatabase, Killer, zAuctionHouseV3, SimpleClans, ajLeaderboards, DeluxeMenus, AuctionGUIPlus, AncientGates, Lottery, PengEventManager, PengUtils, EssentialsSpawn, WorldGuardExtraFlags, ShopGUIPlus, DecentHolograms, Duelos, ShopGUIPlus-SellGUI

### Steps/models to reproduce (if known)

When I activate PlotSquared's 'settings.yml' auto-clear to automatically delete plots from inactive players.

### ImanitySpigot3 version

2022.05.3 BUILD 4

2022.05.3 BUILD 3 - It doesn't have this problem.

### Agreements

- [X] I am running the latest version of ImanitySpigot3 available from
- [X] I have searched for and ensured there isn't already an open issue regarding this.
- [X] I have compared the issue inbetween TacoSpigot 1.8.8 and ImanitySpigot3 And only ImanitySpigot3 occured this issue.

### Other

Video link: https://youtu.be/5qAOPbqutGg

Watch this video that will show that when deleting the plot some chunks can be buggy.

Besides whenever I paste a schematic of the FAWE this problem happens.

**Comment by Weark** [comment]
https://i.imgur.com/T0oFOtk.png

This happens when I load a schematic and paste it with the command //paste

**Comment by Weark** [comment]
![javaw_GHYJpwp4WI](https://user-images.githubusercontent.com/17710865/184462985-80f2bb45-b7e9-470c-9fe0-508ae3da9401.png)

**Comment by Weark** [comment]
I think my problem might be related to #133 

![javaw_pSWffgUqlM](https://user-images.githubusercontent.com/17710865/184463123-b1c6ec91-51dd-4170-86d0-40b7815db8ef.gif)

**Comment by Weark** [comment]
To date, the problem has not been fixed.
Every time I use FAWE the lights are buggy.
I paste a schematic with FAWE and everything goes dark.

**Comment by LeeGodSRC** [comment]
So what exactly is your issue? is it lighting issue or crashing issue?

**Comment by LeeGodSRC** [comment]
Crash reason explained:
FAWE attempts to get chunk at a location that is not loaded, which not safe and FAWE suppose to have a check for it.
But due to ImanitySpigot3 after **2022.05.3 BUILD 4** having a security fallback to let the thread wait until main thread generates the chunk, and also FAWE main thread joins the thread, results in deadlock.

To fix it, there is 3 ways.
1. Don't use FAWE.
2. Build a FAWE fork.
3. Made security fallback to work without the needs for main thread to execute. (Which I 99% wouldn't consider it since async operation for unloaded chunk is not supported at all and may result in performance loose to fix it.)


```
[01:16:26] [ERROR]: ------------------------------
[01:16:26] [ERROR]: Current Thread: Server thread
[01:16:26] [ERROR]: 	PID: 37 | Suspended: false | Native: false | State: RUNNABLE
[01:16:26] [ERROR]: 	Stack:
[01:16:26] [ERROR]: 		java.base@11.0.15/java.lang.Thread.yield(Native Method)
[01:16:26] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinPool.awaitQuiescence(ForkJoinPool.java:2990)
[01:16:26] [ERROR]: 		com.boydti.fawe.example.DefaultFaweQueueMap.next(DefaultFaweQueueMap.java:182)
[01:16:26] [ERROR]: 		com.boydti.fawe.example.MappedFaweQueue.next(MappedFaweQueue.java:220)
[01:16:26] [ERROR]: 		com.boydti.fawe.util.SetQueue$1.run(SetQueue.java:165)
[01:16:26] [ERROR]: 		org.bukkit.craftbukkit.v1_8_R3.scheduler.CraftTask.run(SourceFile:59)
[01:16:26] [ERROR]: 		org.bukkit.craftbukkit.v1_8_R3.scheduler.CraftScheduler.mainThreadHeartbeat(SourceFile:410)
[01:16:26] [ERROR]: 		net.minecraft.server.v1_8_R3.MinecraftServer.B(SourceFile:1026)
[01:16:26] [ERROR]: 		net.minecraft.server.v1_8_R3.DedicatedServer.B(SourceFile:501)
[01:16:26] [ERROR]: 		net.minecraft.server.v1_8_R3.MinecraftServer.A(SourceFile:938)
[01:16:26] [ERROR]: 		net.minecraft.server.v1_8_R3.MinecraftServer.run(SourceFile:828)
[01:16:26] [ERROR]: 		java.base@11.0.15/java.lang.Thread.run(Thread.java:829)
[01:16:26] [ERROR]: ------------------------------
```

```
[01:16:27] [ERROR]: ------------------------------
[01:16:27] [ERROR]: Current Thread: ForkJoinPool-8-worker-3
[01:16:27] [ERROR]: 	PID: 332 | Suspended: false | Native: false | State: WAITING
[01:16:27] [ERROR]: 	Stack:
[01:16:27] [ERROR]: 		java.base@11.0.15/jdk.internal.misc.Unsafe.park(Native Method)
[01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.locks.LockSupport.park(LockSupport.java:194)
[01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.FutureTask.awaitDone(FutureTask.java:447)
[01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.FutureTask.get(FutureTask.java:190)
[01:16:27] [ERROR]: 		net.minecraft.server.v1_8_R3.ChunkProviderServer.getOrCreateChunk(SourceFile:543)
[01:16:27] [ERROR]: 		net.minecraft.server.v1_8_R3.World.getChunkAt(SourceFile:608)
[01:16:27] [ERROR]: 		net.minecraft.server.v1_8_R3.World.getChunkAtWorldCoords(SourceFile:604)
[01:16:27] [ERROR]: 		net.minecraft.server.v1_8_R3.World.getType(SourceFile:1070)
[01:16:27] [ERROR]: 		net.minecraft.server.v1_8_R3.World.getType(SourceFile:1052)
[01:16:27] [ERROR]: 		net.minecraft.server.v1_8_R3.TileEntityChest.b(SourceFile:246)
[01:16:27] [ERROR]: 		net.minecraft.server.v1_8_R3.TileEntityChest.a(SourceFile:228)
[01:16:27] [ERROR]: 		net.minecraft.server.v1_8_R3.TileEntityChest.m(SourceFile:220)
[01:16:27] [ERROR]: 		net.minecraft.server.v1_8_R3.TileEntityChest.y(SourceFile:452)
[01:16:27] [ERROR]: 		com.boydti.fawe.bukkit.v1_8.BukkitChunk_1_8.call(BukkitChunk_1_8.java:278)
[01:16:27] [ERROR]: 		com.boydti.fawe.bukkit.v1_8.BukkitChunk_1_8.call(BukkitChunk_1_8.java:24)
[01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinTask$AdaptedCallable.exec(ForkJoinTask.java:1448)
[01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
[01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinTask.doInvoke(ForkJoinTask.java:408)
[01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:736)
[01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinTask$AdaptedCallable.run(ForkJoinTask.java:1456)
[01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
[01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
[01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
[01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
[01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
[01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
[01:16:27] [ERROR]: ------------------------------
```

**Comment by Weark** [comment]
> Crash reason explained: FAWE attempts to get chunk at a location that is not loaded, which not safe and FAWE suppose to have a check for it. But due to ImanitySpigot3 after **2022.05.3 BUILD 4** having a security fallback to let the thread wait until main thread generates the chunk, and also FAWE main thread joins the thread, results in deadlock.
> 
> To fix it, there is 3 ways.
> 
> 1. Don't use FAWE.
> 2. Build a FAWE fork.
> 3. Made security fallback to work without the needs for main thread to execute. (Which I 99% wouldn't consider it since async operation for unloaded chunk is not supported at all and may result in performance loose to fix it.)
> 
> ```
> [01:16:26] [ERROR]: ------------------------------
> [01:16:26] [ERROR]: Current Thread: Server thread
> [01:16:26] [ERROR]: 	PID: 37 | Suspended: false | Native: false | State: RUNNABLE
> [01:16:26] [ERROR]: 	Stack:
> [01:16:26] [ERROR]: 		java.base@11.0.15/java.lang.Thread.yield(Native Method)
> [01:16:26] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinPool.awaitQuiescence(ForkJoinPool.java:2990)
> [01:16:26] [ERROR]: 		com.boydti.fawe.example.DefaultFaweQueueMap.next(DefaultFaweQueueMap.java:182)
> [01:16:26] [ERROR]: 		com.boydti.fawe.example.MappedFaweQueue.next(MappedFaweQueue.java:220)
> [01:16:26] [ERROR]: 		com.boydti.fawe.util.SetQueue$1.run(SetQueue.java:165)
> [01:16:26] [ERROR]: 		org.bukkit.craftbukkit.v1_8_R3.scheduler.CraftTask.run(SourceFile:59)
> [01:16:26] [ERROR]: 		org.bukkit.craftbukkit.v1_8_R3.scheduler.CraftScheduler.mainThreadHeartbeat(SourceFile:410)
> [01:16:26] [ERROR]: 		net.minecraft.server.v1_8_R3.MinecraftServer.B(SourceFile:1026)
> [01:16:26] [ERROR]: 		net.minecraft.server.v1_8_R3.DedicatedServer.B(SourceFile:501)
> [01:16:26] [ERROR]: 		net.minecraft.server.v1_8_R3.MinecraftServer.A(SourceFile:938)
> [01:16:26] [ERROR]: 		net.minecraft.server.v1_8_R3.MinecraftServer.run(SourceFile:828)
> [01:16:26] [ERROR]: 		java.base@11.0.15/java.lang.Thread.run(Thread.java:829)
> [01:16:26] [ERROR]: ------------------------------
> ```
> 
> ```
> [01:16:27] [ERROR]: ------------------------------
> [01:16:27] [ERROR]: Current Thread: ForkJoinPool-8-worker-3
> [01:16:27] [ERROR]: 	PID: 332 | Suspended: false | Native: false | State: WAITING
> [01:16:27] [ERROR]: 	Stack:
> [01:16:27] [ERROR]: 		java.base@11.0.15/jdk.internal.misc.Unsafe.park(Native Method)
> [01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.locks.LockSupport.park(LockSupport.java:194)
> [01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.FutureTask.awaitDone(FutureTask.java:447)
> [01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.FutureTask.get(FutureTask.java:190)
> [01:16:27] [ERROR]: 		net.minecraft.server.v1_8_R3.ChunkProviderServer.getOrCreateChunk(SourceFile:543)
> [01:16:27] [ERROR]: 		net.minecraft.server.v1_8_R3.World.getChunkAt(SourceFile:608)
> [01:16:27] [ERROR]: 		net.minecraft.server.v1_8_R3.World.getChunkAtWorldCoords(SourceFile:604)
> [01:16:27] [ERROR]: 		net.minecraft.server.v1_8_R3.World.getType(SourceFile:1070)
> [01:16:27] [ERROR]: 		net.minecraft.server.v1_8_R3.World.getType(SourceFile:1052)
> [01:16:27] [ERROR]: 		net.minecraft.server.v1_8_R3.TileEntityChest.b(SourceFile:246)
> [01:16:27] [ERROR]: 		net.minecraft.server.v1_8_R3.TileEntityChest.a(SourceFile:228)
> [01:16:27] [ERROR]: 		net.minecraft.server.v1_8_R3.TileEntityChest.m(SourceFile:220)
> [01:16:27] [ERROR]: 		net.minecraft.server.v1_8_R3.TileEntityChest.y(SourceFile:452)
> [01:16:27] [ERROR]: 		com.boydti.fawe.bukkit.v1_8.BukkitChunk_1_8.call(BukkitChunk_1_8.java:278)
> [01:16:27] [ERROR]: 		com.boydti.fawe.bukkit.v1_8.BukkitChunk_1_8.call(BukkitChunk_1_8.java:24)
> [01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinTask$AdaptedCallable.exec(ForkJoinTask.java:1448)
> [01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
> [01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinTask.doInvoke(ForkJoinTask.java:408)
> [01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:736)
> [01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinTask$AdaptedCallable.run(ForkJoinTask.java:1456)
> [01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
> [01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.FutureTask.run(FutureTask.java:264)
> [01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
> [01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
> [01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
> [01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
> [01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
> [01:16:27] [ERROR]: 		java.base@11.0.15/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
> [01:16:27] [ERROR]: ------------------------------
> ```

Most of the plugins I use depend on FAWE, including PlotSquared which has this problem.

Unfortunately I can't stop using FAWE and I can't make a fork of FAWE either. Aside from the FAWE/PlotSquared team who are co-owned, they no longer support version 1.8.

Can you make an update which I can disable/enable this function that is crashing the server?

**Comment by Weark** [comment]
> I've added something that attempts to fix the problem in **2023.3.1 BUILD 2**, please give it a try. @Weark

Great, it's fixed, but I still see the light issue when using FAWE.

And FAWE now gives an error when used, see this print: https://i.imgur.com/Gnd0Fv5.png

**Comment by LeeGodSRC** [comment]
The error of FAWE seems normal
And the lighting issue has another issue [#133](https://github.com/Imanity-Software/ImanitySpigot3-git/issues/133), feel free to provide more information there.
